### ISSUE TITLE
terra::rast from MASS::kde2d does not work like raster::raster

### ISSUE THREAD

**Issue by JepsonNomad** [rspatial/terra#630]
Hi there - I found the following inconsistency between raster and terra packages. raster allows you to directly convert a list with the elements x = vector, y = vector, z = matrix(len(x),len(y)) to a raster. However, terra does not cooperate in the same way, and when calling to z directly weird results occur, presumably related to the indexing differences among matrices and rasters.

```
## Import packages
library(raster)
library(terra)
library(MASS)
## Generate data
inDat = data.frame("x" = c(1,10,100),
                   "y" = c(50,10,90))
## Run kernel density
kde_example = MASS::kde2d(x = as.vector(inDat$x),
                          y = as.vector(inDat$y),
                          n = c(100,100))
image(kde_example)
## Convert to raster data
r1 = raster::raster(kde_example)
plot(r1) # This plot matches image(kde_example)
r2 = terra::rast(kde_example) # Error: [rast,list] none of the elements of x are a SpatRaster
r3 = terra::rast(kde_example$z) # Generates a result but an unexpected one
plot(r3)
```

**Comment by kadyb** [comment]
This code changes the indexing and returns the same results as `{raster}`:

```r
mat = kde_example$z
mat = apply(t(mat), 2, rev) # or mat = t(mat[, ncol(mat):1L])
r3 = terra::rast(mat)
plot(r3)
```

**Comment by JepsonNomad** [comment]
Thanks for the workaround @kadyb - this solves the problem in my example above but in a larger problem that I can't make reproducible the result is basically a bunch of noise (seems more complicated than just a transposition). 

The result of `image(kde)` is as expected:

<img width="432" alt="Screen Shot 2022-05-07 at 11 13 48 AM" src="https://user-images.githubusercontent.com/33075946/167266762-2506bf3c-e626-40f3-b6cd-0f64d4c6c009.png">

But with `plot(r3)` (`r3` generated according to kadyb's code above but with full kde rather than example) is odd:

![weirdResult](https://user-images.githubusercontent.com/33075946/167266236-9e0bfa30-498a-43fe-8c53-4d90d1f5fcd9.png)

It seems like whatever the pipeline that connects the list structure of `kde` to `raster::raster()` should be implemented in terra.

Also possibly relevant:
```
> r3
class       : SpatRaster 
dimensions  : 2225, 2000, 1  (nrow, ncol, nlyr)
resolution  : 1, 1  (x, y)
extent      : 0, 2000, 0, 2225  (xmin, xmax, ymin, ymax)
coord. ref. :  
source      : memory 
name        :         lyr.1 
min value   : 6.405617e-135 
max value   :  2.600911e-13 
> str(r3)
Formal class 'SpatRaster' [package "terra"] with 1 slot
  ..@ ptr:Reference class 'Rcpp_SpatRaster' [package "terra"] with 20 fields
  .. ..$ depth    : num 0
  .. ..$ extent   :Reference class 'Rcpp_SpatExtent' [package "terra"] with 2 fields
  .. .. ..$ valid : logi TRUE
  .. .. ..$ vector: num [1:4] 0 2000 0 2225
  .. .. ..and 28 methods, of which 14 are  possibly relevant:
  .. .. ..  align, as.points, ceil, compare, deepcopy, finalize, floor,
  .. .. ..  initialize, intersect, round, sample, sampleRandom,
  .. .. ..  sampleRegular, union
  .. ..$ filenames: chr ""
  .. ..$ hasRange : logi TRUE
  .. ..$ hasTime  : logi FALSE
  .. ..$ hasUnit  : logi FALSE
  .. ..$ hasValues: logi TRUE
  .. ..$ inMemory : logi TRUE
  .. ..$ messages :Reference class 'Rcpp_SpatMessages' [package "terra"] with 2 fields
  .. .. ..$ has_error  : logi FALSE
  .. .. ..$ has_warning: logi FALSE
  .. .. ..and 18 methods, of which 4 are  possibly relevant:
  .. .. ..  finalize, getError, getWarnings, initialize
  .. ..$ names    : chr "lyr.1"
  .. ..$ origin   : num [1:2] 0 0
  .. ..$ range_max: num 2.6e-13
  .. ..$ range_min: num 6.41e-135
  .. ..$ res      : num [1:2] 1 1
  .. ..$ rgb      : logi FALSE
  .. ..$ rgbtype  : chr ""
  .. ..$ time     : num 0
  .. ..$ timestep : chr "seconds"
  .. ..$ units    : chr ""
  .. ..$ valueType: int 0
  .. ..and 215 methods, of which 201 are  possibly relevant:
  .. ..  addSource, adjacent, adjacentMat, aggregate, align, apply,
  .. ..  area_by_value, arith_numb, arith_rast, as_lines, as_points,
  .. ..  as_polygons, atan2, bilinearValues, boundaries, buffer,
  .. ..  and many many more [remainder of list removed by me]
```

**Comment by kadyb** [comment]
Maybe creating raster from coordinates will work:

```r
df = expand.grid(x = kde_example$x, y = kde_example$y, KEEP.OUT.ATTRS = FALSE)
df$z = as.vector(kde_example$z)
r3 = rast(df)
plot(r3)
```

**Comment by JepsonNomad** [comment]
Ok, I tried that as well and it returned the same result. However, I realized that the values generated by MASS::kde2d are _extremely_ small - see the output for `print(r3)` - and it occurred to me that the way terra is generating a color ramp in `plot()` suggests it's treating values as factors rather than continuous numeric. So, I tried multiplying the df$z by 1e11 and sure enough, it produced a result that follows the spatial pattern of intensity that I would expect. So we are zeroing in on an actual problem: what is the finest precision allowable by terra?
```
df = expand.grid(x = kde$x, y = kde$y, KEEP.OUT.ATTRS = FALSE)
df$z = as.vector(kde$z)*(1e11)
r3 = rast(df)
plot(r3)
```
![expectedresult](https://user-images.githubusercontent.com/33075946/167269634-45485801-bef1-4d6b-8851-110594d5cd92.png)

As far as my original problem goes, I'd call this one solved. However, the problem of terra treating a raster grid of reallllyyyyy small values as factors seems like a real issue that should be addressed.
